File: leetcode/js/relative-sort-array.js
/**
 * @param {number[]} arr1
 * @param {number[]} arr2
 * @return {number[]}
 */
var relativeSortArray = function (arr1, arr2) {
  const set = new Set(arr2);

  const contain = [];
  const uncontain = [];

  for (const num of arr1) {
    if (set.has(num)) {
      contain.push(num);
    } else {
      uncontain.push(num);
    }
  }

  contain.sort((a, b) => arr2.indexOf(a) - arr2.indexOf(b));
  uncontain.sort((a, b) => a - b);

  return contain.concat(uncontain);
};

File: leetcode/rust/relative-sort-array.rs
impl Solution {
    pub fn relative_sort_array(arr1: Vec<i32>, arr2: Vec<i32>) -> Vec<i32> {
        let mut contain: Vec<i32> = arr1
            .iter()
            .filter(|&num| arr2.contains(num))
            .map(|num| *num)
            .collect();
        let mut uncontain: Vec<i32> = arr1
            .iter()
            .filter(|&num| !arr2.contains(num))
            .map(|num| *num)
            .collect();

        contain.sort_unstable_by(|a, b| {
            arr2.iter()
                .position(|num| a == num)
                .unwrap()
                .cmp(&arr2.iter().position(|num| b == num).unwrap())
        });
        uncontain.sort_unstable();

        vec![contain, uncontain].concat()
    }
}
